block-layer/content-addressable-archives.md

```diff
@@ -0,0 +1,150 @@
+# Specification: Content Addressable aRchives (CAR / .car)
+
+**Status: Draft**
+
+## Summary
+
+The CAR format (Content Addressable aRchives) can be used to store content addressable objects in the form of IPLD block data as a sequence of bytes; typically in a file with a `.car` filename extension.
+
+The CAR format is intended as a serialized representation of any IPLD DAG (graph) as the concatenation of its blocks, plus a header that describes the graphs in the file (via root CIDs). The requirement for the blocks in a CAR to form coherent DAGs is not strict, so the CAR format may also be used to store any arbitrary list of IPLD blocks.
+
+In addition to the binary block data, storage overhead for the CAR format consists of:
+
+ * A header block encoded as [DAG-CBOR](codecs/dag-cbor.md) containing the format version and an array of root CIDs
+ * A CID for each block preceding its binary data
+ * A compressed integer prefixing each block, including the header block, indicating the total length of that block including CID
+
+This diagram shows how IPLD blocks, their root CID, and a header combine to form a CAR.
+
+![Content Addressable aRchive Diagram](content-addressable-archives.png)
+
+## Format Description
+
+The CAR format comprises a sequence of length-prefixed IPLD block data, where the first block in the CAR is the Header encoded as CBOR, and the remaining blocks form the Data component of the CAR and are each additionally prefixed with their CIDs. The length prefix of each block in a CAR is encoded as a "varint"&mdash;an unsigned [LEB128](https://en.wikipedia.org/wiki/LEB128) integer. This integer specifies the number of remaining bytes for that block entry&mdash;excluding the bytes used to encode the integer, but including the CID for non-header blocks.
+
+```
+|--------- Header --------| |---------------------------------- Data -----------------------------------|
+
+[ varint | dag-cbor block ] [ varint | CID | block ] [ varint | CID | block ] [ varint | CID | block ] …
+```
+
+### Header
+
+The first bytes of the CAR format hold a varint, this unsigned integer specifies the number of bytes beyond the varint itself that contain the _Header_ block. This Header block is a byte array DAG-CBOR (CBOR with tag 42 for CIDs) encoded object holding the version number and array of roots. As an [IPLD Schema](../schemas/):
```

Comment:
So, we _could_ just make this entire thing recursive. That is, the entire CAR could be:

```
[ varint | CID | BLOCK ] *
```

Where the _first_ block is always the header object.

Reply:
That would make a nice symmetry. Unfortunately the version information is in the header block and there's no CID before that block, so if we had a v2 that introduced a CID for the header block then our parsers would need to be able to figure out whether the bytes after the initial varint are a CID or the start of a CBOR block (`map['version':x,...`). Is the difference between the start of a CID and the start of a CBOR map distinct enough for such branching?

Reply:
My thinking was to make that version _1_. If we also required that the root block list all root blocks, the CAR would actually be a single DAG and would be really easy to validate.

Reply:
I like that idea. Reading the header would be more work (as you would also need to read its CID), but that should be negligible compared to making parsing easier as it's less of a special case.